### PR TITLE
Deepnet prototype

### PR DESCRIPTION
**Pull request by mmerce**
@charleslparker I added the deepnet.py file which has a predict method based on your code.  In its present state, it accepts as first argument a single input data in the usual dictionary form. This should be transformed to the expected array format you use internally by using the fill_array method. In this method we should do all the logic to transform missings or text/items to what you need to process.

I've tried to use the code as is with the iris.csv in 1-clicks and an input data with no missings, to test the easiest case. The output is a vector, which is not normalized. I assume it refers to the probabilities of the classes in the objective field. Unfortunately, they don't match the probabilites I get if I do the same predictions in tropo (even after normalization). I'd appreciate if you could have a look and help me find what is missing or wrong.

You'll see that some code has been commented in the predict function that uses the .tree attribute. I believe the deepnet I tested does not use this attribute, so I don't think this is the problem right now. However, I'd also need to understand what's this attribute going to contain. Folloiwng your code, I saw some scikit-learn calls and objects that I'm not sure I need to add here. Do we also need scikit-learn for deepnet to make predictions?

By the way, you mentioned a conditional import of numpy, but I see numpy used in both the modules I believe you used. If I'm not wrong, the conditional import is in scipy, right? 

Thanks

### CODE REVIEW

**Comment by mmerce** [comment]
By the way, the PR is only so that we can comment on the code.

**Comment by charleslparker** [comment]
> I've tried to use the code as is with the iris.csv in 1-clicks and an input data with no missings, to test the easiest case. The output is a vector, which is not normalized. 

That's a red flag.  The output of the network should be the output of a softmax layer, which should always sum to 1.  This should be true regardless of the input values you feed in.

> However, I'd also need to understand what's this attribute going to contain. 

It's essentially a JSON list of lists.  The lists specify a tree-based embedding for the input data that's used to preprocess it before it gets to the rest of the network.  I can go over it with you in as much detail as you like.

> Folloiwng your code, I saw some scikit-learn calls and objects that I'm not sure I need to add here.   Do we also need scikit-learn for deepnet to make predictions?

No.  I use scikit to generate some of these objects, but they're converted to JSON when the structure goes to wintermute.

> You'll see that some code has been commented in the predict function that uses the .tree attribute. I believe the deepnet I tested does not use this attribute, so I don't think this is the problem right now. 

Yes, you're right.

> By the way, you mentioned a conditional import of numpy, but I see numpy used in both the modules 

If you look at where numpy is used in the `math_ops` module, there are three places.  Two are just checking to see if an input data structure is a numpy ndarray, and the last one is the sum for `sum_and_normalize`, which is only used for that soon-to-be-obsolete ensemble code I mentioned.  These things should be easily removable.

>  I'd appreciate if you could have a look and help me find what is missing or wrong.

At least part of the problem here is something that I missed telling you:  Before the data gets to the network, it goes through this function:

https://github.com/bigmlcom/laminar/blob/next/laminar/data/preprocess.py#L137

which standardizes numeric variables and converts categoricals to one-hot binaries.  The `specs` in that function are the list of maps under the `preprocess` key in the network.

So this would happen *before* the predict method (or, equivalently, at the beginning thereof).

I'm not sure this is the entire problem - the fact that you're not getting a proper distribution as output is still somewhat concerning - but it is part of the problem.

**Comment by mmerce** [comment]
Ok, so I'll start by addiing this preprocessing code. However, the preprocessing function you mention

https://github.com/bigmlcom/laminar/blob/next/laminar/data/preprocess.py#L137

uses numpy

 column = np.asarray(column, dtype=np.float32)

so does it really make sense still having this conditional on numpy?

**Comment by charleslparker** [comment]
> so does it really make sense still having this conditional on numpy?

Everything that's done in those functions is easy to do without numpy (it's just element-wise math).  I'm agnostic towards whether numpy is used or not; I thought it was very important to us not to have the dependency, which I understood and which is possible without too much work.  But I also understand if we want to avoid that work and have the extra speed.

**Comment by mmerce** [comment]
> Everything that's done in those functions is easy to do without numpy (it's just element-wise math).

But it's not done anywhere in your code, right? I can add that once we got this working.

**Comment by charleslparker** [comment]
> But it's not done anywhere in your code, right? I can add that once we got this working.

I'm somewhat certain the necessary functions are already in `math_ops`, but you'd (or I) would have to find and use them.

**Comment by mmerce** [comment]
Ok, I'll have a look once we got something predicting with np.

**Comment by mmerce** [comment]
Still a bit lost. The transformations use this variable specs, which contains (for numerics) something like:
spec:  {'stdev': 0.0, 'type': 'numeric', 'mean': 1.0}

so, these stdev and mean refer to the ones in the field, right?

**Comment by charleslparker** [comment]
You should get something like (this is the clojure syntax, so like this but in JSON):

```
{:index 0, :mean 5.84333, :type numeric, :stdev 0.8253}
```

for each speck in the list_of_specs, so if you have all numeric features:

```
output_vector = []
for spec in specs:
    output_vector.append((input_vector[spec['index']] - spec['mean']) / spec['stdev'])
```

**Comment by mmerce** [comment]
closing this PR and opening another one against the bigmlcom branch
